cf-custom-resources/lib/wkld-cert-validator.js

```diff
@@ -0,0 +1,881 @@
+  let id;
+  return async () => {
+    if (!id) {
+      id = await hostedZoneIDByName(`${appName}.${domainName}`);
+    }
+    return id;
+  };
+};
+
+const rootHostedZoneIDContext = () => {
+  let id;
+  return async () => {
+    if (!id) {
+      id = await hostedZoneIDByName(`${domainName}`);
+    }
+    return id;
+  };
+};
+
+let hostedZoneID = {
+  app: appHostedZoneIDContext(),
+  root: rootHostedZoneIDContext(),
+};
+
+/**
+ * Upload a CloudFormation response object to S3.
+ *
+ * @param {object} event the Lambda event payload received by the handler function
+ * @param {object} context the Lambda context received by the handler function
+ * @param {string} responseStatus the response status, either 'SUCCESS' or 'FAILED'
+ * @param {string} physicalResourceId CloudFormation physical resource ID
+ * @param {object} [responseData] arbitrary response data object
+ * @param {string} [reason] reason for failure, if any, to convey to the user
+ * @returns {Promise} Promise that is resolved on success, or rejected on connection error or HTTP error response
+ */
+function report(
+  event,
+  context,
+  responseStatus,
+  physicalResourceId,
+  responseData,
+  reason
+) {
+  return new Promise((resolve, reject) => {
+    const https = require("https");
+    const { URL } = require("url");
+
+    let reasonWithLogInfo = `${reason} (Log: ${context.logGroupName}/${context.logStreamName})`;
+    let responseBody = JSON.stringify({
+      Status: responseStatus,
+      Reason: reasonWithLogInfo,
+      PhysicalResourceId: physicalResourceId || context.logStreamName,
+      StackId: event.StackId,
+      RequestId: event.RequestId,
+      LogicalResourceId: event.LogicalResourceId,
+      Data: responseData,
+    });
+
+    const parsedUrl = new URL(event.ResponseURL);
+    const options = {
+      hostname: parsedUrl.hostname,
+      port: 443,
+      path: parsedUrl.pathname + parsedUrl.search,
+      method: "PUT",
+      headers: {
+        "Content-Type": "",
+        "Content-Length": responseBody.length,
+      },
+    };
+
+    https
+      .request(options)
+      .on("error", reject)
+      .on("response", (res) => {
+        res.resume();
+        if (res.statusCode >= 400) {
+          reject(new Error(`Error ${res.statusCode}: ${res.statusMessage}`));
+        } else {
+          resolve();
+        }
+      })
+      .end(responseBody, "utf8");
+  });
+}
+
+exports.handler = async function (event, context) {
+  // Destruct resource properties into local variables.
+  const props = event.ResourceProperties;
+  let { LoadBalancerDNS: loadBalancerDNS } = props;
+  const aliases = new Set(props.Aliases);
+
+  // Initialize global variables.
+  envHostedZoneID = props.EnvHostedZoneId;
+  envName = props.EnvName;
+  appName = props.AppName;
+  serviceName = props.ServiceName;
+  domainName = props.DomainName;
+  rootDNSRole = props.RootDNSRole;
+  cloudFrontCert = props.IsCloudFrontCertificate;
+  certificateDomain = cloudFrontCert
```

Comment:
Instead of a `bool`, what do you think of receiving an enum that accepts either `CloudFront` or `NLB`.

Reply:
Would the refactor trigger an unnecessary update to the existing NLB custom resource / existing static site custom resource though?

Reply:
There'll be a super fast update to update the custom resource action input. but it'll quit instantly because there's no substantial update for DNS.